README.md
# php-qrencode

PHP qrencode extension, binding to [libqrencode](https://github.com/fukuchi/libqrencode).

This PHP extension is essentially the work found at http://hirokawa.netflowers.jp/entry/4900/ (I can not find the author's name).

That site doesn't seem to exist anymore, so I am putting the code here for the benefit of everyone. I've also made some improvements (see CHANGELOG.md for details).

C programming is not my forté, so if you have any suggested improvements, please fork and send a pull request.


## Building

First, you need to compile and install [libqrencode](https://github.com/fukuchi/libqrencode).
Please see that package's README for instructions.

Then, compile php-qrencode:

```
cd ../php-qrencode
phpize
./configure
make
sudo make install
```

Finally, add the following to the end of your `php.ini` file:

```
extension=qrencode.so
```

## Usage

```php
array qrencode( string $data, int $ecc, int $version, int $mode );
```

`$data` is the information to encode in the QR Code (e.g. URL, vCard, phone number, etc.).

`$ecc` is one of the error correction level constants defined below.

`$version` (integer in the range 1-40) defines the overall block size of the generated image (block size = (version * 4) + 17).  If you pass in more data than can fit in the requested size, the function will increase the size as required so this, in effect, defines the _minimum_ size of the resulting QR Code.

`$mode` is one of the storage modes defined below.

The returned two-dimensional array defines the on/off state for each pixel "block" in the resulting QR code.

To turn this data into an image, you will need to use GD functions, and loop through the array, building the image as you go.  e.g.:

```php
$data = qrencode( 'Hello world!', QR_ECLEVEL_L, 1, QR_MODE_8 );

$size = count( $data );
$im = imagecreate( $size, $size );

$color_bg = imagecolorallocate( $im, 255, 255, 255 );
$color_fg = imagecolorallocate( $im, 0, 0, 0 );

imagefill( $im, 0, 0, $color_bg );

for( $y = 0; $y < $size; $y++ ) {
	for( $x = 0; $x < $size; $x++ ) {
		if ( $data[$y][$x] ) {
			imagesetpixel( $im, $x, $y, $color_fg );
		}
	}
}

imagepng( $im );
```


## PHP Constants

### Storage Modes

- `QR_MODE_NUM` - numeric only, max. 7,089 characters, 0-9

- `QR_MODE_AN` - alphanumeric, max. 4,296 characters, 0-9, A-Z (uppercase only), space, $, %, *, +, -, ., /, :

- `QR_MODE_8` - binary/byte, max. 2,953 characters (8-bit bytes) or 23624 bits

- `QR_MODE_KANJI` - Kanji/Kana, max. 1,817 characters

### Error Correction Levels

- `QR_ECLEVEL_L` - Low (approximately 7% of codewords can be restored)

- `QR_ECLEVEL_M` - Medium (15%)

- `QR_ECLEVEL_Q` - Quartile (25%)

- `QR_ECLEVEL_H` - High (30%)


-----

"QR Code" is a registered trademark of Denso Wave Incorporated.

CHANGELOG.md
# Changelog


### Version 0.0.3 - 18-Sep-2012

- added support for compilation under PHP 5.4
- tested with current libqrencode version (3.3.1)


### Version 0.0.2

- this is the code originally taken from http://hirokawa.netflowers.jp/entry/4900/


php_qrencode.h
/*
 * php-qrencode
 *
 * ORIGINAL AUTHOR, LICENSE AND COPYRIGHT OWNER NOT KNOWN
 * CODE ORIGINALLY FROM http://hirokawa.netflowers.jp/entry/4900/
 * BUT NO LONGER AVAILABLE.
 *
 * THE FOLLOWING COPYRIGHT NOTICE IS ENTIRELY DEPENDENT ON
 * THE WISHES OF THE ORIGINAL AUTHOR.
 *
 * Copyright (c) 2012 Colin Viebrock <colin@viebrock.ca>
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject
 * to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

#ifndef PHP_QRENCODE_H
#define PHP_QRENCODE_H

#ifdef  __cplusplus
extern "C" {
#endif

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <php.h>

#ifdef HAVE_QRENCODE

#include <php_ini.h>
#include <SAPI.h>
#include <ext/standard/info.h>
#include <Zend/zend_extensions.h>
#ifdef  __cplusplus
} // extern "C"
#endif
#include <qrencode.h>
#ifdef  __cplusplus
extern "C" {
#endif

extern zend_module_entry qrencode_module_entry;
#define phpext_qrencode_ptr &qrencode_module_entry

#ifdef PHP_WIN32
#define PHP_QRENCODE_API __declspec(dllexport)
#else
#define PHP_QRENCODE_API
#endif

PHP_MINIT_FUNCTION(qrencode);
PHP_MSHUTDOWN_FUNCTION(qrencode);
PHP_RINIT_FUNCTION(qrencode);
PHP_RSHUTDOWN_FUNCTION(qrencode);
PHP_MINFO_FUNCTION(qrencode);

#ifdef ZTS
#include "TSRM.h"
#endif

#define FREE_RESOURCE(resource) zend_list_delete(Z_LVAL_P(resource))

#define PROP_GET_LONG(name)    Z_LVAL_P(zend_read_property(_this_ce, _this_zval, #name, strlen(#name), 1 TSRMLS_CC))
#define PROP_SET_LONG(name, l) zend_update_property_long(_this_ce, _this_zval, #name, strlen(#name), l TSRMLS_CC)

#define PROP_GET_DOUBLE(name)    Z_DVAL_P(zend_read_property(_this_ce, _this_zval, #name, strlen(#name), 1 TSRMLS_CC))
#define PROP_SET_DOUBLE(name, d) zend_update_property_double(_this_ce, _this_zval, #name, strlen(#name), d TSRMLS_CC)

#define PROP_GET_STRING(name)    Z_STRVAL_P(zend_read_property(_this_ce, _this_zval, #name, strlen(#name), 1 TSRMLS_CC))
#define PROP_GET_STRLEN(name)    Z_STRLEN_P(zend_read_property(_this_ce, _this_zval, #name, strlen(#name), 1 TSRMLS_CC))
#define PROP_SET_STRING(name, s) zend_update_property_string(_this_ce, _this_zval, #name, strlen(#name), s TSRMLS_CC)
#define PROP_SET_STRINGL(name, s, l) zend_update_property_string(_this_ce, _this_zval, #name, strlen(#name), s, l TSRMLS_CC)


PHP_FUNCTION(qrencode);
#ifdef  __cplusplus
} // extern "C"
#endif

#endif /* PHP_HAVE_QRENCODE */

#endif /* PHP_QRENCODE_H */


/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */


qrencode.c
/*
 * php-qrencode
 *
 * ORIGINAL AUTHOR, LICENSE AND COPYRIGHT OWNER NOT KNOWN
 * CODE ORIGINALLY FROM http://hirokawa.netflowers.jp/entry/4900/
 * BUT NO LONGER AVAILABLE.
 *
 * THE FOLLOWING COPYRIGHT NOTICE IS ENTIRELY DEPENDENT ON
 * THE WISHES OF THE ORIGINAL AUTHOR.
 *
 * Copyright (c) 2012 Colin Viebrock <colin@viebrock.ca>
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject
 * to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

#include "php_qrencode.h"

#if HAVE_QRENCODE

/* {{{ qrencode_functions[] */
#if (PHP_MAJOR_VERSION == 5 && PHP_MINOR_VERSION < 4) || (PHP_MAJOR_VERSION < 5)
function_entry qrencode_functions[] = {
#else
static zend_function_entry qrencode_functions[] = {
#endif
	PHP_FE(qrencode, NULL)
	{ NULL, NULL, NULL }
};
/* }}} */


/* {{{ qrencode_module_entry
 */
zend_module_entry qrencode_module_entry = {
	STANDARD_MODULE_HEADER,
	"qrencode",
	qrencode_functions,
	PHP_MINIT(qrencode),     /* Replace with NULL if there is nothing to do at php startup   */
	PHP_MSHUTDOWN(qrencode), /* Replace with NULL if there is nothing to do at php shutdown  */
	PHP_RINIT(qrencode),     /* Replace with NULL if there is nothing to do at request start */
	PHP_RSHUTDOWN(qrencode), /* Replace with NULL if there is nothing to do at request end   */
	PHP_MINFO(qrencode),
	"0.0.3",
	STANDARD_MODULE_PROPERTIES
};
/* }}} */

#ifdef COMPILE_DL_QRENCODE
ZEND_GET_MODULE(qrencode)
#endif


/* {{{ PHP_MINIT_FUNCTION */
PHP_MINIT_FUNCTION(qrencode)
{
	REGISTER_LONG_CONSTANT("QR_MODE_NUM", QR_MODE_NUM, CONST_PERSISTENT | CONST_CS);
	REGISTER_LONG_CONSTANT("QR_MODE_AN", QR_MODE_AN, CONST_PERSISTENT | CONST_CS);
	REGISTER_LONG_CONSTANT("QR_MODE_8", QR_MODE_8, CONST_PERSISTENT | CONST_CS);
	REGISTER_LONG_CONSTANT("QR_MODE_KANJI", QR_MODE_KANJI, CONST_PERSISTENT | CONST_CS);
	REGISTER_LONG_CONSTANT("QR_ECLEVEL_L", QR_ECLEVEL_L, CONST_PERSISTENT | CONST_CS);
	REGISTER_LONG_CONSTANT("QR_ECLEVEL_M", QR_ECLEVEL_M, CONST_PERSISTENT | CONST_CS);
	REGISTER_LONG_CONSTANT("QR_ECLEVEL_Q", QR_ECLEVEL_Q, CONST_PERSISTENT | CONST_CS);
	REGISTER_LONG_CONSTANT("QR_ECLEVEL_H", QR_ECLEVEL_H, CONST_PERSISTENT | CONST_CS);

	/* add your stuff here */

	return SUCCESS;
}
/* }}} */


/* {{{ PHP_MSHUTDOWN_FUNCTION */
PHP_MSHUTDOWN_FUNCTION(qrencode)
{

	/* add your stuff here */

	return SUCCESS;
}
/* }}} */


/* {{{ PHP_RINIT_FUNCTION */
PHP_RINIT_FUNCTION(qrencode)
{
	/* add your stuff here */

	return SUCCESS;
}
/* }}} */


/* {{{ PHP_RSHUTDOWN_FUNCTION */
PHP_RSHUTDOWN_FUNCTION(qrencode)
{
	/* add your stuff here */

	return SUCCESS;
}
/* }}} */


/* {{{ PHP_MINFO_FUNCTION */
PHP_MINFO_FUNCTION(qrencode)
{
	php_info_print_box_start(0);
	php_printf("<p>QRCode Extension</p>\n");
	php_printf("<p>Version 0.0.3</p>\n");
	php_info_print_box_end();
	/* add your stuff here */

}
/* }}} */


/* {{{ proto array qrencode(string text,int version,int level,int mode)
   */
PHP_FUNCTION(qrencode)
{
	const char * text = NULL;
	int text_len = 0;
	long version = 0;
	long level = 0;
	long mode = 0;

	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "slll", &text, &text_len, &version, &level, &mode) == FAILURE) {
		return;
	}
	array_init(return_value);

	do {
//		QRinput *input;
		QRcode *code;
		unsigned char *p;
		int x,y;

//		input = QRinput_new();
//		QRinput_append(input,mode,text_len,text);
		// as per ttp://hirokawa.netflowers.jp/entry/4900/#comment_51714
		//code = QRcode_encodeInput(input,version,level);
		//code = QRcode_encodeInput(input);

		//code = QRcode_encodeString(text, version, level, mode);

		code = QRcode_encodeString(text, version, level, mode, 1);


		if(code == NULL) {
			RETURN_FALSE;
		}

		array_init(return_value);

		p = code->data;
		for(y=0;y<code->width;y++){

			zval *new_array;
			MAKE_STD_ZVAL(new_array);
			array_init(new_array);

			for(x=0;x<code->width;x++){
				add_next_index_long(new_array, (*p &1)?1:0);
				p++;
			}
			add_next_index_zval(return_value, new_array);
		}

//		QRinput_free(input);
		QRcode_free(code);
	} while (0);
}
/* }}} qrencode */

#endif /* HAVE_QRENCODE */


/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */
